crypto/verificationhelper/qrcode.go

```diff
@@ -0,0 +1,93 @@
+package verificationhelper
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+
+	"go.mau.fi/util/random"
+
+	"maunium.net/go/mautrix/id"
+)
+
+var (
+	ErrInvalidQRCodeHeader  = errors.New("invalid QR code header")
+	ErrUnknownQRCodeVersion = errors.New("invalid QR code version")
+	ErrInvalidQRCodeMode    = errors.New("invalid QR code mode")
+)
+
+type QRCodeMode byte
+
+const (
+	QRCodeModeCrossSigning                    QRCodeMode = 0x00
+	QRCodeModeSelfVerifyingMasterKeyTrusted   QRCodeMode = 0x01
+	QRCodeModeSelfVerifyingMasterKeyUntrusted QRCodeMode = 0x02
+)
+
+type QRCode struct {
+	Mode          QRCodeMode
+	TransactionID id.VerificationTransactionID
+	Key1, Key2    [32]byte
+	SharedSecret  []byte
+}
+
+func NewQRCode(mode QRCodeMode, txnID id.VerificationTransactionID, key1, key2 [32]byte) *QRCode {
+	return &QRCode{
+		Mode:          mode,
+		TransactionID: txnID,
+		Key1:          key1,
+		Key2:          key2,
+		SharedSecret:  random.Bytes(32),
+	}
+}
+
+// NewQRCodeFromBytes parses the bytes from a QR code scan as defined in
+// [Section 11.12.2.4.1] of the Spec.
+//
+// [Section 11.12.2.4.1]: https://spec.matrix.org/v1.9/client-server-api/#qr-code-format
+func NewQRCodeFromBytes(data []byte) (*QRCode, error) {
+	data = bytes.TrimPrefix(data, []byte("QR-Code:"))
```

Comment:
What's this trim for 🤔

Reply:
For some reason I was getting that from the QR code scanning program that I was using. I guess we can just rely on it not ever being present.